File: Array/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if(n==0) {
            return 0;
        }
        int out = 0;

        int rightMax = prices[n-1];

        for(int i=n-2; i >=0;i--) {
            rightMax = Math.max(rightMax, prices[i]);
            out = Math.max(out, rightMax - prices[i]);
        }
        return out;

    }
}

File: Array/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        maxPrice = prices[-1]
        for i in range(len(prices)-2, -1, -1):
            maxPrice = max(maxPrice, prices[i])
            maxProfit = max(maxProfit, maxPrice - prices[i])
        return maxProfit